File: internal/usecases/dto/notice.go
package dto

import "strconv"

// =====================
// NoticeTypeForm は通知種別のフォーム DTO である。
type NoticeTypeForm struct {
	ID        int64        `json:"id"`
	Name      string       `json:"name"`
	CreatedAt *string      `json:"createdAt,omitempty"`
	UpdatedAt *string      `json:"updatedAt,omitempty"`
	DeletedAt *string      `json:"deletedAt,omitempty"`
	Notices   []NoticeForm `json:"notices,omitempty"`
}

// =====================
// NoticeForm は通知のフォーム DTO である。
type NoticeForm struct {
	ID         int64           `json:"id"`
	TypeID     int64           `json:"typeId"`
	QuestionID *string         `json:"questionId,omitempty"`
	Content    *string         `json:"content,omitempty"`
	DisplayDue *string         `json:"displayDue,omitempty"`
	CreatedAt  *string         `json:"createdAt,omitempty"`
	UpdatedAt  *string         `json:"updatedAt,omitempty"`
	DeletedAt  *string         `json:"deletedAt,omitempty"`
	NoticeType *NoticeTypeForm `json:"noticeType,omitempty"`
	Question   *QuestionForm   `json:"question,omitempty"`
}

// QuestionIDInt64 は QuestionID 文字列を int64 に変換する。nil または空文字の場合は -1 を返す。
//
// return:
//   - int64: 質問 ID（未設定・変換失敗時は -1）
func (f NoticeForm) QuestionIDInt64() int64 {
	if f.QuestionID == nil || *f.QuestionID == "" {
		return -1
	}
	if val, err := strconv.ParseInt(*f.QuestionID, 10, 64); err == nil {
		return val
	}
	return -1
}

File: static/js/models/Notice.js
/**
 * Notice model class.
 */

import { NoticeType } from '/static/js/models/NoticeType.js';
import { Question } from '/static/js/models/Question.js';
import { formToApiJson } from '/static/js/models/helpers.js';

export class Notice {
  /**
   * @param {number} id
   * @param {number} typeId
   * @param {number|null} questionId
   * @param {string|null} content
   * @param {string|null} displayDue ISO8601 or null
   * @param {string|null} createdAt
   * @param {string|null} updatedAt
   * @param {NoticeType|null} [noticeType]
   * @param {Question|null} [question]
   */
  constructor(id, typeId, questionId, content, displayDue, createdAt = null, updatedAt = null, noticeType = null, question = null) {
    this.id = id;
    this.typeId = typeId;
    this.questionId = questionId;
    this.content = content;
    this.displayDue = displayDue ? new Date(displayDue) : null;
    this.createdAt = createdAt ? new Date(createdAt) : null;
    this.updatedAt = updatedAt ? new Date(updatedAt) : null;
    this.noticeType = noticeType;
    this.question = question;
  }

  /** @param {Object} json @returns {Notice} */
  static fromJSON(json) {
    return new Notice(
      json.id,
      json.typeId,
      json.questionId ?? null,
      json.content ?? null,
      json.displayDue ?? null,
      json.createdAt ?? null,
      json.updatedAt ?? null,
      json.noticeType ? NoticeType.fromJSON(json.noticeType) : null,
      json.question ? Question.fromJSON(json.question) : null,
    );
  }

  /** @returns {Object} */
  static toModel(form) {
    return formToApiJson(form);
  }
}
